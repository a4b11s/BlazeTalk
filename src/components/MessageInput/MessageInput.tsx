import React, { FormEvent, useState } from 'react';

import FileInput from '../FileInput/FileInput';
import ImageList from '../ImageList/ImageList';

import { Box, CircularProgress, IconButton, TextField } from '@mui/material';
import SendIcon from '@mui/icons-material/Send';

interface IProps {
	onSend: (message: string, imageArray?: Array<File>) => void;
	isSending?: boolean;
}

interface IImage {
	blobUrl: string;
	file: File;
	uid: symbol;
}

const MessageInput = (props: IProps) => {
	const { onSend, isSending } = props;
	const [messageText, setMessageText] = useState('');
	const [messageImages, setMessageImages] = useState<
		Array<IImage> | undefined
	>();

	const onMessageTextChange = (event: any) => {
		setMessageText(event.target.value);
	};

	const onFileUploaded = (FileList: FileList) => {
		const FileArray = Array.from(FileList);
		const imageArray: Array<IImage> = [];

		FileArray.forEach((file) => {
			let blob = new Blob([file], {
				type: 'application/image',
			});
			imageArray.push({
				blobUrl: URL.createObjectURL(blob),
				file,
				uid: Symbol(file.name),
			});
		});

		setMessageImages((prevState) => {
			if (!prevState) return imageArray;
			return [...prevState, ...imageArray];
		});
	};

	const onImageDelete = (uid: symbol) => {
		setMessageImages((prevState) => {
			if (!prevState) return undefined;

			const deletingImageIndex = prevState?.findIndex(
				(image) => image.uid === uid
			);

			return [
				...prevState.slice(0, deletingImageIndex),
				...prevState.slice(deletingImageIndex + 1),
			];
		});
	};

	const onSendBtnClick = (e: FormEvent<HTMLFormElement>) => {
		e.preventDefault();

		if (messageImages) {
			const messageFile: Array<File> = messageImages?.map((image) => image.file);
			onSend(messageText, messageFile);
			setMessageImages(undefined);
		} else {
			onSend(messageText);
		}

		setMessageText('');
	};

	return (
		<Box px="2%">
			<ImageList onDeleteButtonClick={onImageDelete} images={messageImages} />
			<Box py={1}>
				<form style={{ height: '100%', display: 'flex' }} onSubmit={onSendBtnClick}>
					<TextField
						autoFocus
						sx={{ flexGrow: 1 }}
						value={messageText}
						label="Message"
						variant="outlined"
						onChange={onMessageTextChange}
					/>
					<FileInput multiple accept="image/*" onFileUploaded={onFileUploaded} />
					<IconButton
						sx={{ position: 'relative' }}
						disabled={messageText.trim() === '' && messageImages === undefined}
						type="submit"
						aria-label="send"
						color="secondary"
					>
						{isSending && (
							<CircularProgress size={50} sx={{ position: 'absolute' }} />
						)}
						<SendIcon fontSize="large" />
					</IconButton>
				</form>
			</Box>
		</Box>
	);
};

export default MessageInput;
