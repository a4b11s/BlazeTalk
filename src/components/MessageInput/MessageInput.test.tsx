import React from 'react';

import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MessageInput from './MessageInput';

describe('MessageInput component', () => {
    test('should be render', () => {
        render(<MessageInput onSend={() => {
        }}/>);
        expect(screen.getByRole('textbox')).toBeInTheDocument();
        expect(screen.getByRole('button')).toBeInTheDocument();
    });
    describe('send button should be disable and onSend not called', () => {
        test('if message empty', () => {
            render(<MessageInput onSend={() => {
            }}/>);
            expect(screen.getByRole('textbox')).toBeInTheDocument();
            expect(screen.getByRole('button')).toHaveAttribute('disabled');
        });
        test('if message white space', async () => {
            render(<MessageInput onSend={() => {
            }}/>);
            await waitFor(() => {
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.type(screen.getByRole('textbox'), ' ');
            });

            expect(screen.getByRole('button')).toHaveAttribute('disabled');
        });
        test('if message white tab', async () => {
            render(<MessageInput onSend={() => {
            }}/>);
            await waitFor(() => {
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.type(screen.getByRole('textbox'), '	');
            });
            expect(screen.getByRole('button')).toHaveAttribute('disabled');
        });
    });
    describe('should be call onSend', () => {
        test('if press send button', async () => {
            const onSend = jest.fn();
            render(<MessageInput onSend={onSend}/>);
            await waitFor(() => {
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.type(screen.getByRole('textbox'), 'test-text');
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.click(screen.getByRole('button'));
            });
            expect(onSend).toHaveBeenCalledTimes(1);
        });
        test('if press enter key', async () => {
            const onSend = jest.fn();
            render(<MessageInput onSend={onSend}/>);
            await waitFor(() => {
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.type(screen.getByRole('textbox'), 'test-text');
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.click(screen.getByRole('textbox'));
                // eslint-disable-next-line testing-library/no-wait-for-side-effects
                userEvent.keyboard('{ENTER}');
            });
            expect(onSend).toHaveBeenCalledTimes(1);
        });
    });
    test('should to match snapshot with correct props', () => {
        const view = render(<MessageInput onSend={() => {
        }}/>);
        expect(view).toMatchSnapshot();
    });
});
