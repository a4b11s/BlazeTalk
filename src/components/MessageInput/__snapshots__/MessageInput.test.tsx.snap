// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput component should to match snapshot with correct props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-dhk8m9"
      >
        <form
          style="height: 100%; display: flex;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-te7hn0-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r6:"
              id=":r6:-label"
            >
              Message
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r6:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Message
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            aria-label="send"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="SendIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
              />
            </svg>
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-dhk8m9"
    >
      <form
        style="height: 100%; display: flex;"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-te7hn0-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r6:"
            id=":r6:-label"
          >
            Message
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r6:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Message
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          aria-label="send"
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="SendIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
